crates/loro-internal/src/state/tree_state.rs

```diff
@@ -26,10 +33,462 @@ use super::ContainerState;
 #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, EnumAsInner, Serialize)]
 pub enum TreeParentId {
     Node(TreeID),
+    // We use `Unexist` as the old parent of a new node created
+    // so we can infer the retreat internal diff is `Uncreate`
     Unexist,
     Deleted,
-    /// parent is root
-    None,
+    Root,
+}
+
+impl From<Option<TreeID>> for TreeParentId {
+    fn from(id: Option<TreeID>) -> Self {
+        match id {
+            Some(id) => {
+                if TreeID::is_deleted_root(&id) {
+                    TreeParentId::Deleted
+                } else {
+                    TreeParentId::Node(id)
+                }
+            }
+            None => TreeParentId::Root,
+        }
+    }
+}
+
+#[derive(Clone)]
+enum NodeChildren {
+    Vec(Vec<(NodePosition, TreeID)>),
+    BTree(btree::ChildTree),
+}
+
+impl Default for NodeChildren {
+    fn default() -> Self {
+        NodeChildren::Vec(vec![])
+    }
+}
+
+impl NodeChildren {
+    fn get_index_by_child_id(&self, target: &TreeID) -> Option<usize> {
+        match self {
+            NodeChildren::Vec(v) => v.iter().position(|(_, id)| id == target),
+            NodeChildren::BTree(btree) => btree.id_to_index(target),
+        }
+    }
+
+    fn get_node_position_at(&self, pos: usize) -> &NodePosition {
+        match self {
+            NodeChildren::Vec(v) => &v[pos].0,
+            NodeChildren::BTree(btree) => &btree.get_elem_at(pos).unwrap().pos,
+        }
+    }
+
+    fn get_elem_at(&self, pos: usize) -> Option<(&NodePosition, &TreeID)> {
+        match self {
+            NodeChildren::Vec(v) => v.get(pos).map(|x| (&x.0, &x.1)),
+            NodeChildren::BTree(btree) => btree.get_elem_at(pos).map(|x| (x.pos.as_ref(), &x.id)),
+        }
+    }
+
+    #[cfg(not(feature = "tree_jitter"))]
+    fn generate_fi_at(&self, pos: usize, target: &TreeID) -> FractionalIndexGenResult {
```

Comment:
If feature = 'tree_jitter' is on, this function will not be called, right?

Reply:
At the bottom of this file, I add a mod named `jitter`.